airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/DestinationRunner.kt

```diff
@@ -0,0 +1,232 @@
+package io.airbyte.cdk.write
+
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.math.max
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+/**
+ * Runs a DestinationTask state machine, either from an arbitrary initial task
+ * or by wrapping a StandardDestination in the default state machine entry point.
+ */
+class DestinationRunner(
+    private val initialTask: DestinationTask
```

Comment:
wild thought: it would be cool for this to be List<Task>, since then we could very easily have our "page 1"
```kotlin
DestinationRunner(listOf(
  DefaultSetup(destination),
  DefaultAccumulateRecords(destination),
  ForEachStream {
    WhenAllComplete {
      typeAndDedupe()
    }
  }
)).run()
```

(... there's a whole bunch of problems with this naively though. Like presumably the setup() call needs to pass info to typeAndDedupe().)

Reply:
I LOVE this.

It could actually eliminate the need for the standard interface, since you can just express it in the "page 1". Even better, it eliminates that sticky threshold between "my job still kind of conforms to the interface so I'll hack it in" and "I guess I have to use the state machine now."

Ie, you're always using the state machine, but we just give you really expressive semantics that are just as intuitive as an interface, so if you need to make slight adjustments it's obvious how.
 
Also, the tasks can easily share state w/o being part of a class, so that's not an issue.

Reply:
Okay, the obvious tradeoff here is that it requires lifting the task sequencing semantics out of the tasks themselves, which complicates them a little (since less is done in code, more in objects). However, I think it addresses the issue of A) not wanting the tasks to have too much control over the process BUT B) not wanting too much bespoke stuff in the runner.

Looks like this

```
```
taskList(
  DefaultSetup(),
  ForEachStream { stream -> taskList(
    DefaultOpenStream(stream),
    Replicated(nAccumulators) { index ->
      UntilComplete(MessageQueue.getConsumer(stream, index)) {
        YieldWhenNull { consumer ->
          DefaultAccumulateRecords(consumer)
        }
      }
    }.forEachOutput { batch ->
      Incrementing("$stream:batch") {
        DefaultProcessBatch(batch)
      }
    }.forEachOutput { batch ->
      ReprocessUntilNull { batch ->
        MaybeMerging({it.mergeable}) {
          DefaultProcessBatch(batch)
        }
      }.finally {
        Decrementing("$stream:batch") {
          ExactlyOnce("$stream:eos") {
            WhenAllReleased("$stream:batch") {
              DefaultCloseStream(stream)
            }
          }
        }
      }
    }
 )},
 DefaultTeardown()
)
```

Reply:
(that `ReprocessUntilNull...finally Decrementing...` could just be another `taskList(ReprocessUntilNull, Decrementing)`)